src/test/assembly/asm/hexagon-types.rs

```diff
@@ -0,0 +1,136 @@
+
+#![feature(no_core, lang_items, rustc_attrs, repr_simd)]
+#![crate_type = "rlib"]
+#![no_core]
+#![allow(asm_sub_register, non_camel_case_types)]
+
+#[rustc_builtin_macro]
+macro_rules! asm {
+    () => {};
+}
+#[rustc_builtin_macro]
+macro_rules! stringify {
+    () => {};
+}
+
+#[lang = "sized"]
+trait Sized {}
+#[lang = "copy"]
+trait Copy {}
+
+type ptr_8 = *const i8;
+type ptr_16 = *const i16;
+type ptr_32 = *const i32;
+type ptr_mut_32 = *mut i32;
+
+impl Copy for i8 {}
+impl Copy for i16 {}
+impl Copy for i32 {}
+impl Copy for f32 {}
+impl Copy for ptr_32 {}
+impl Copy for ptr_16 {}
+impl Copy for ptr_8 {}
+extern "C" {
+    fn extern_func();
+    static extern_static: u8;
+}
+
+macro_rules! check_transfer {
+    ($func:ident $ty:ident $class:ident) => {
+        #[no_mangle]
+        pub unsafe fn $func(x: $ty) -> $ty {
+            // Hack to avoid function merging
+            extern "Rust" {
+                fn dont_merge(s: &str);
+            }
+            dont_merge(stringify!($func));
+
+            let y;
+            asm!("{} = {}", out($class) y, in($class) x);
+            y
+        }
+    };
+}
+
+
+// CHECK-LABEL: gpr_transfer_sym:
+// CHECK: InlineAsm Start
+// CHECK: r0 = #extern_static
+// CHECK: InlineAsm End
+#[no_mangle]
+pub unsafe fn gpr_transfer_sym() {
+            // Hack to avoid function merging
+            extern "Rust" {
+                fn dont_merge(s: &str);
+            }
+            dont_merge(stringify!($func));
+    asm!("r0 = #{}", sym extern_static);
+}
+
+// CHECK-LABEL: packet:
+// CHECK: InlineAsm Start
+// CHECK: {
+// CHECK:   r{{[0-9]+}} = r0
+// CHECK:   memw(r1) = r{{[0-9]+}}
+// CHECK: }
+// CHECK: InlineAsm End
+#[no_mangle]
+pub unsafe fn packet() {
+    let val = 1024;
+    asm!("{{
+{} = r0
+memw(r1) = {}
+}}", out(reg) _, in(reg) &val);
+}
+
+
+// CHECK-LABEL: ptr_8:
+// CHECK: InlineAsm Start
+// CHECK: r{{[0-9]+}} = r{{[0-9]+}}
+// CHECK: InlineAsm End
+check_transfer!(ptr_8 ptr_8 reg);
+
+// CHECK-LABEL: ptr_16:
+// CHECK: InlineAsm Start
+// CHECK: r{{[0-9]+}} = r{{[0-9]+}}
+// CHECK: InlineAsm End
+check_transfer!(ptr_16 ptr_16 reg);
+
+// CHECK-LABEL: ptr_32:
+// CHECK: InlineAsm Start
+// CHECK: r{{[0-9]+}} = r{{[0-9]+}}
+// CHECK: InlineAsm End
+check_transfer!(ptr_32 ptr_32 reg);
+
+// CHECK-LABEL: float_32:
+// CHECK: InlineAsm Start
+// CHECK: r{{[0-9]+}} = r{{[0-9]+}}
+// CHECK: InlineAsm End
+check_transfer!(float_32 f32 reg);
+
+// CHECK-LABEL: gpr_32:
+// CHECK: InlineAsm Start
+// CHECK: r{{[0-9]+}} = r{{[0-9]+}}
+// CHECK: InlineAsm End
+check_transfer!(gpr_32 i32 reg);
+
+// CHECK-LABEL: gpr_8:
+// CHECK: InlineAsm Start
+// CHECK: r{{[0-9]+}} = r{{[0-9]+}}
+// CHECK: InlineAsm End
+check_transfer!(gpr_8 i8 reg);
+
+// CHECK-LABEL: gpr_16:
+// CHECK: InlineAsm Start
+// CHECK: r{{[0-9]+}} = r{{[0-9]+}}
+// CHECK: InlineAsm End
+check_transfer!(gpr_16 i16 reg);
+
+// CHECK-LABEL: pred_8:
+// CHECK: InlineAsm Start
+// CHECK: r{{[0-9]+}} = r{{[0-9]+}}
```

Comment:
Shouldn't this be checking that the register is `p*` instead of `r*`? In fact I don't think LLVM exposes a way to allocate a predicate register via inline asm, so we will probably have to remove the preg class.

Reply:
You're right.  That's also why I took out the vector regs.  They're missing from upstream llvm.  I'll see about making sure they get contributed upstream and then either cherry pick them onto the rust llvm or wait for them to come via 11.0.0.  Then I'll revisit the inline asm regs here.

For now GPRs should suffice.